File: 백준/Gold/9663. N－Queen/N－Queen.py
# 220915 9663 N-Queen

# 정답코드

import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**5)

N = int(input())
# 체스판
board = [[0 for _ in range(N)] for _ in range(N)]

# 열, 대각선1, 대각선2
visited_col = [0 for _ in range(N)]
visited_diagonal_1 = [0 for _ in range(2 * N)]
visited_diagonal_2 = [0 for _ in range(2 * N)]

# 엔퀸 개수
cnt = 0


def n_queen(row, N):
    global cnt

    # 깊이 도달 시 
    if row == N:
        cnt += 1
        return

    for col in range(N):
        # 미방문
        if visited_col[col] == 0 and visited_diagonal_1[row + col] == 0 and visited_diagonal_2[row - col + N] == 0:

            # 방문처리
            visited_col[col] = 1
            visited_diagonal_1[row + col] = 1
            visited_diagonal_2[row - col + N] = 1   

            n_queen(row + 1, N)

            # 백트래킹
            visited_col[col] = 0
            visited_diagonal_1[row + col] = 0
            visited_diagonal_2[row - col + N] = 0  


n_queen(0, N)

print(cnt)

File: 백준/Gold/9663. N－Queen/N－Queen.js
const fs = require("fs");
const inputData = fs.readFileSync("/dev/stdin").toString();

const n = parseInt(inputData);


function findQueens(n) {
  // 이곳에 작성합니다.
  const visited_col = new Array(n).fill(0);
  const visited_diagonal_1 = new Array(2 * n).fill(0);
  const visited_diagonal_2 = new Array(2 * n).fill(0);

  let answer = 0;

  const n_queen = (row, n) => {
    if (row === n) {
      answer += 1;
      return;
    }

    for (let col = 0; col < n; col++) {
      if (
        visited_col[col] == 0 &&
        visited_diagonal_1[row + col] == 0 &&
        visited_diagonal_2[row - col + n] == 0
      ) {
        visited_col[col] = 1;
        visited_diagonal_1[row + col] = 1;
        visited_diagonal_2[row - col + n] = 1;

        n_queen(row + 1, n);

        visited_col[col] = 0;
        visited_diagonal_1[row + col] = 0;
        visited_diagonal_2[row - col + n] = 0;
      }
    }
  };

  n_queen(0, n);
  return answer;
}

console.log(findQueens(n));
